src/Telemetry/Library/TelemetryManager.cs

```diff
@@ -0,0 +1,444 @@
+
+                    if (tearDownLogManager)
+                    {
+                        LogManager.Teardown();
+                    }
+                },
+                initializingOrTearingDown: true);
+
+        /// <summary>
+        /// Logs all public properties of the object into an event, respecting the following rules:
+        /// - Exception objects will be partially serialized to Json respecting the Configuration.ExceptionLoggingOptions.
+        ///   - The name of the event will be "Quantum_{AppId}_.Exception"
+        ///   - We only serialize the exception full type name, and based on the ExceptionLoggingOptions we can serialize:
+        ///     - TargetSite (the signature of the method that raised the exception)
+        ///     - SanitizedStackTrace (the stack trace without the file paths)
+        ///   - Other properties are not captured as they could contain customer data or PII.
+        /// - Null objects won't be logged.
+        /// - Value types will not get logged.
+        /// - Other object types will be logged in an event named $"Quantum_{AppId}_{object type name}".
+        ///   - Properties with null values will not be logged.
+        ///   - Properties of Value types will be converted to values accepted by Aria.
+        ///   - Properties with Nullable types will be converted their corresponding non-nullable types accepted by Aria.
+        ///   - Properties of Exception types will be partially serialized to Json respecting the Configuration.ExceptionLoggingOptions.
+        ///   - Properties of Enum types will be converted to strings.
+        ///   - Properties with [PiiData] attribute will be marked as PII and will be hashed.
+        ///   - Properties with [SerializeJson] attribute will be serialized as Json.
+        ///   - All other property types won't be logged.
+        /// </summary>
+        public static void LogObject(object obj) =>
+            CheckAndRunSafe(() =>
+            {
+                if (obj == null || obj is ValueType)
+                {
+                    return;
+                }
+
+                if (obj is EventProperties eventProp)
+                {
+                    LogEvent(eventProp);
+                    return;
+                }
+
+                EventProperties eventProperties = new();
+                string? eventName = null;
+
+                // We don't log exception fields as they can potentially contain customer data
+                if (obj is Exception exception)
+                {
+                    obj = exception.ToExceptionLogRecord(Configuration.ExceptionLoggingOptions);
+                    eventName = $"{Configuration.EventNamePrefix}_{Configuration.AppId}_Exception";
+                }
+
+                var type = obj.GetType();
+                eventProperties.Name = eventName ?? $"{Configuration.EventNamePrefix}_{Configuration.AppId}_{type.Name}";
+                var properties = ReflectionCache.GetProperties(type);

```

Comment:
Are we worried that reflection is very slow? Since one of the goals was to make this fast so it doesn't add a significant overhead to short processes like qsfmt, I'm wondering if this defeats the purpose.

Reply:
Yes, in general reflection is slow, but not that slow if only used once, hence why I created a cache.
IMO, the convenience of using reflection outweighs the verbose code of logging each property line by line.
This makes it a lot easier for developers to instrument applications by definition the events schemas as strongly typed records.
With everything else that goes on with the instrumentation (like uploading data, out-of-process communication, etc.) the cost of a one-time reflection should be neglectable. 
If we really want to optimize this, I would rather use a code generation tool to "reflect" at compile time rather than asking the developer to log property by property.